File: sprint_5/misc/bst/is_bst.c
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>

#define MAX_NODES 20

typedef struct BTNode {
  int value;
  struct BTNode *left;
  struct BTNode *right;
} BTNode;

// TODO: implement visualisation
void print_binary_tree(BTNode *root);

// Time: O(n) Space: O(h). Best for simple trees.
bool is_BST_recursive(BTNode *root, int low, int high) {
  if (root == NULL)
    return true;
  if (root->value <= low || root->value >= high)
    return false;
  return is_BST_recursive(root->left, low, root->value) &&
         is_BST_recursive(root->right, root->value, high);
}

typedef struct {
  BTNode *node;
  int low;
  int high;
} BTNodeComparator;

// Time: O(n) Space: O(h). Best for deeper trees.
bool is_BST_iterative(BTNode *root) {
  if (root == NULL)
    return true;
  int i = 0;
  BTNodeComparator stack[MAX_NODES];
  stack[i++] =
      (BTNodeComparator){.node = root, .low = INT_MIN, .high = INT_MAX};
  while (i > 0) {
    BTNodeComparator current = stack[--i];
    BTNode *node = current.node;
    if (node->value <= current.low || node->value >= current.high) {
      return false;
    };
    if (node->left != NULL) {
      BTNodeComparator ndc_left = {
          .node = node->left, .low = current.low, .high = node->value};
      stack[i++] = ndc_left;
    }
    if (node->right != NULL) {
      BTNodeComparator ndc_right = {
          .node = node->right, .low = node->value, .high = current.high};
      stack[i++] = ndc_right;
    }
  }
  return true;
}

// Time: O(n) Space: O(h). Generally better approach.
bool is_BST_inorder(BTNode *root) {
  if (root == NULL)
    return true;
  int i = 0;
  int prev = INT_MIN;
  BTNode *stack[MAX_NODES];
  while (root != NULL || i > 0) {
    while (root != NULL) {
      stack[i++] = root;
      root = root->left;
    }
    BTNode *current = stack[--i];
    if (current->value <= prev) {
      return false;
    }
    prev = current->value;
    root = current->right;
  }
  return true;
}
void confirm_result(bool got, bool expected, const char *test_num) {
  if (got == expected) {
    printf("test %s passed!\n", test_num);
  } else {
    printf("test %s failed!\n", test_num);
  }
}

int main(void) {
  // Define valid BST here
  BTNode node7 = {.value = 20, .left = NULL, .right = NULL};
  BTNode node6 = {.value = 15, .left = NULL, .right = NULL};
  BTNode node5 = {.value = 16, .left = &node6, .right = NULL};
  BTNode node4 = {.value = 9, .left = NULL, .right = NULL};
  BTNode node3 = {.value = 4, .left = NULL, .right = NULL};
  BTNode node2 = {.value = 18, .left = &node5, .right = &node7};
  BTNode node1 = {.value = 8, .left = &node3, .right = &node4};
  BTNode root = {.value = 12, .left = &node1, .right = &node2};

  // Test recursive implementation
  bool validated_bst = is_BST_recursive(&root, INT_MIN, INT_MAX);
  confirm_result(validated_bst, true, "is BST recursive positive");
  node4.value = 148;
  validated_bst = is_BST_recursive(&root, INT_MIN, INT_MAX);
  confirm_result(validated_bst, false, "is BST recursive negative");
  node4.value = 9;

  // Test iterative implementation
  validated_bst = is_BST_iterative(&root);
  confirm_result(validated_bst, true, "is BST iterative positive");
  node4.value = 148;
  validated_bst = is_BST_iterative(&root);
  confirm_result(validated_bst, false, "is BST iterative negative");
  node4.value = 9;

  // Test inorder implementation
  validated_bst = is_BST_inorder(&root);
  confirm_result(validated_bst, true, "is BST iterative positive");
  node4.value = 148;
  validated_bst = is_BST_inorder(&root);
  confirm_result(validated_bst, false, "is BST iterative negative");
  node4.value = 9;
  return 0;
}

File: sprint_5/misc/bst/is_bst.py
from typing import Callable, Self


class Node:
    def __init__(self, value: int, left: Self | None = None, right: Self | None = None):
        self.value = value
        self.right = right
        self.left = left


def is_bst_recursive(root: Node) -> bool:
    def validate(node: Node, low=float("-inf"), high=float("inf")) -> bool:
        if not node:
            return True
        if not (low < node.value < high):
            return False
        return validate(node.left, low, node.value) and validate(node.right, node.value, high)

    return validate(root)


def is_bst_iterative(root: Node) -> bool:
    stack = [(root, float("-inf"), float("inf"))]
    while stack:
        node, low, high = stack.pop()
        if not (low < node.value < high):
            return False
        if node.left:
            stack.append((node.left, low, node.value))
        if node.right:
            stack.append((node.right, node.value, high))
    return True


def is_bst_inorder(root: Node) -> bool:
    prev = float("-inf")
    stack = []
    node = root
    while node or stack:
        while node:
            stack.append(node)
            node = node.left
        node = stack.pop()
        if node.value <= prev:
            return False
        prev = node.value
        node = node.right
    return True


def test(implementations: tuple[Callable]):
    node1 = Node(1, None, None)
    node2 = Node(4, None, None)
    node3 = Node(3, node1, node2)
    node4 = Node(8, None, None)
    node5 = Node(5, node3, node4)

    for i, impl in enumerate(implementations, 1):
        assert impl(node5)
        print(f"test {i}.1 passed!")
        node2.value = 5
        assert not impl(node5)
        print(f"test {i}.2 passed!")
        node2.value = 4


if __name__ == "__main__":
    test((is_bst_recursive, is_bst_iterative, is_bst_inorder))
